modules/ingest-common/src/main/java/org/opensearch/ingest/common/CommunityIDProcessor.java

```diff
@@ -0,0 +1,654 @@
+        return sourceIPField;
+    }
+
+    public String getSourcePortField() {
+        return sourcePortField;
+    }
+
+    public String getDestinationIPField() {
+        return destinationIPField;
+    }
+
+    public String getDestinationPortField() {
+        return destinationPortField;
+    }
+
+    public String getIANAProtocolNumberField() {
+        return ianaProtocolNumberField;
+    }
+
+    public String getProtocolField() {
+        return protocolField;
+    }
+
+    public String getIcmpTypeField() {
+        return icmpTypeField;
+    }
+
+    public String getIcmpCodeField() {
+        return icmpCodeField;
+    }
+
+    public int getSeed() {
+        return seed;
+    }
+
+    public String getTargetField() {
+        return targetField;
+    }
+
+    public boolean isIgnoreMissing() {
+        return ignoreMissing;
+    }
+
+    @Override
+    public IngestDocument execute(IngestDocument document) {
+        // resolve protocol firstly
+        Protocol protocol = resolveProtocol(document);
+        // exit quietly if protocol cannot be resolved and ignore_missing is true
+        if (protocol == null) {
+            return document;
+        }
+
+        // resolve ip secondly, exit quietly if either source ip or destination ip cannot be resolved and ignore_missing is true
+        byte[] sourceIPByteArray = resolveIP(document, sourceIPField);
+        if (sourceIPByteArray == null) {
+            return document;
+        }
+        byte[] destIPByteArray = resolveIP(document, destinationIPField);
+        if (destIPByteArray == null) {
+            return document;
+        }
+        // source ip and destination ip must have same format, either ipv4 or ipv6
+        if (sourceIPByteArray.length != destIPByteArray.length) {
+            throw new IllegalArgumentException("source ip and destination ip must have same format");
+        }
+
+        // resolve source port and destination port for transport protocols,
+        // exit quietly if either source port or destination port is null nor empty
+        Integer sourcePort = null;
+        Integer destinationPort = null;
+        final boolean isTransportProtocol = Protocol.isTransportProtocol(protocol.getProtocolCode());
+        if (isTransportProtocol) {
+            sourcePort = resolvePort(document, sourcePortField);
+            if (sourcePort == null) {
+                return document;
+            }
+
+            destinationPort = resolvePort(document, destinationPortField);
+            if (destinationPort == null) {
+                return document;
+            }
+        }
+
+        // resolve ICMP message type and code, support both ipv4 and ipv6
+        // set source port to icmp type, and set dest port to icmp code, so that we can have a generic way to handle
+        // all protocols
+        boolean isOneway = true;
+        final boolean isICMPProtocol = Protocol.ICMP.getProtocolCode() == protocol.getProtocolCode()
```

Comment:
The beauty of enumerations:

```suggestion
        final boolean isICMPProtocol = Protocol.ICMP == protocol || Protocol.ICMP_V6 == protocol;
```